src/jsonpatcherproxy.js

```diff
@@ -172,24 +221,55 @@ var JSONPatcherProxy = (function() {
     for (let key in root) {
       if (root.hasOwnProperty(key)) {
         if (typeof root[key] === 'object') {
-          const distPath =
+          const destinationPropKey =
             path + '/' + JSONPatcherProxy.escapePathComponent(key);
-          root[key] = this.generateProxyAtPath(root[key], distPath);
-          this._proxifyObjectTreeRecursively(root[key], distPath);
+          root[key] = this.generateProxyAtPath(root[key], destinationPropKey);
+          this._proxifyObjectTreeRecursively(root[key], destinationPropKey);
         }
       }
     }
     return this.generateProxyAtPath(root, path);
   };
+  JSONPatcherProxy.prototype._resetCachedProxiesPaths = function(root, path) {
+    /* deleting array elements could render other array elements paths incorrect, 
+    this function fixes all incorrect paths efficiently */
+    for (let key in root) {
+      if(!root.hasOwnProperty(key)) continue;
+      const subObject = root[key];
+      if (
+        subObject && subObject._isProxified
+      ) {
+        const destinationPropKey = path + '/' + JSONPatcherProxy.escapePathComponent(key);
+
+        // get the proxy instance (an instance is {proxy, originalObject})
+        var revokableProxyInstance = this.proxifiedObjectsMap.get(subObject);
+        if (revokableProxyInstance) {
+          // get the un-proxified originalObject, we need because `set` trap passes the original object as target
+          const originalObject = revokableProxyInstance.originalObject;
+          // update its path
+          this.objectsPathsMap.set(originalObject, destinationPropKey);
+        }
+        /* Even though we didn't find the proxy in the proxies-cache,
+         We need to continue updating its descendants' paths,
+         because we might come at this point in recursive calls inside `set` trap.
+         Because we cache the proxy (revokableInstance) to proxies map only **after** the revokableInstance is created.
+         And `set` trap can call this function before revokableInstance is created.
+         This means we can come across totally valid proxies that are not cached in the our cache.
+         We continue traversing them because we know 100% they're proxies (they have _isProxified = true).
```

Comment:
> How do we know they have _isProxified = true?
 
Because [that's the condition ](https://github.com/Palindrom/JSONPatcherProxy/blob/Fix-nested-array-deletion/src/jsonpatcherproxy.js#L240)to reaching this line.

> What if we will change the order of proxification to proxify parent first, then child? Would we be able to just reset proxies that has revokableInstance?

The Order of proxification to proxify is already parent first. It is just the parent proxy isnt' always cached before the child proxy is cached. 

Caching happens after the whole tree is proxified.